File: translations/go/apexkit/realtime.go
package apexkit

import (
	"encoding/json"
	"fmt"
	"log"
	"strings"
	"sync"
	"time"

	"github.com/gorilla/websocket"
)

type SubscriptionFilter struct {
	CollectionID int                    `json:"collection_id,omitempty"`
	RecordID     int                    `json:"record_id,omitempty"`
	EventType    string                 `json:"event_type,omitempty"`
	DataFilter   map[string]interface{} `json:"filter,omitempty"`
	Channel      string                 `json:"channel,omitempty"`
	CustomEvent  string                 `json:"custom_event,omitempty"`
}

type ApexKitRealtimeWSClient struct {
	URL               string
	Token             string
	Conn              *websocket.Conn
	mu                sync.Mutex
	listeners         []func(msg interface{})
	isConnected       bool
	currentFilter     *SubscriptionFilter
	reconnectInterval time.Duration
}

func NewRealtimeWSClient(baseURL, token string) *ApexKitRealtimeWSClient {
	wsURL := strings.Replace(baseURL, "http", "ws", 1) + "/ws"
	return &ApexKitRealtimeWSClient{
		URL:               wsURL,
		Token:             token,
		reconnectInterval: 3 * time.Second,
	}
}

func (c *ApexKitRealtimeWSClient) Connect() {
	go func() {
		for {
			err := c.connect()
			if err != nil {
				log.Printf("[ApexKit] Connection failed: %v. Retrying in %v...", err, c.reconnectInterval)
				time.Sleep(c.reconnectInterval)
				continue
			}
			break
		}
	}()
}

func (c *ApexKitRealtimeWSClient) connect() error {
	dialer := websocket.DefaultDialer
	conn, _, err := dialer.Dial(c.URL, nil)
	if err != nil {
		return err
	}

	c.mu.Lock()
	c.Conn = conn
	c.isConnected = true
	c.mu.Unlock()

	log.Println("[ApexKit] Realtime Connected")

	if c.currentFilter != nil {
		c.Subscribe(*c.currentFilter)
	}

	for {
		_, message, err := conn.ReadMessage()
		if err != nil {
			c.mu.Lock()
			c.isConnected = false
			c.mu.Unlock()
			return err
		}

		var msg interface{}
		if err := json.Unmarshal(message, &msg); err != nil {
			if string(message) == "Pong" {
				continue
			}
			log.Printf("WS Parse Error: %v", err)
			continue
		}

		c.notify(msg)
	}
}

func (c *ApexKitRealtimeWSClient) Subscribe(filter SubscriptionFilter) error {
	c.mu.Lock()
	c.currentFilter = &filter
	conn := c.Conn
	connected := c.isConnected
	c.mu.Unlock()

	if !connected || conn == nil {
		return fmt.Errorf("not connected")
	}

	msg := map[string]interface{}{
		"type":    "Subscribe",
		"payload": filter,
	}
	return conn.WriteJSON(msg)
}

func (c *ApexKitRealtimeWSClient) SendSignal(channel, eventName string, data interface{}) error {
	c.mu.Lock()
	conn := c.Conn
	connected := c.isConnected
	c.mu.Unlock()

	if !connected || conn == nil {
		return fmt.Errorf("not connected")
	}

	msg := map[string]interface{}{
		"type": "Signal",
		"payload": map[string]interface{}{
			"channel": channel,
			"event":   eventName,
			"data":    data,
		},
	}
	return conn.WriteJSON(msg)
}

func (c *ApexKitRealtimeWSClient) OnEvent(listener func(msg interface{})) func() {
	c.mu.Lock()
	c.listeners = append(c.listeners, listener)
	c.mu.Unlock()

	return func() {
		c.mu.Lock()
		defer c.mu.Unlock()
		for i, l := range c.listeners {
			if fmt.Sprintf("%p", l) == fmt.Sprintf("%p", listener) {
				c.listeners = append(c.listeners[:i], c.listeners[i+1:]...)
				break
			}
		}
	}
}

func (c *ApexKitRealtimeWSClient) notify(msg interface{}) {
	c.mu.Lock()
	defer c.mu.Unlock()
	for _, l := range c.listeners {
		l(msg)
	}
}

File: translations/python/apexkit/realtime.py
import json
import asyncio
import websockets
import uuid
from typing import Optional, Dict, Any, Callable, List
import requests

class ApexKitRealtimeWSClient:
    def __init__(self, url: str, token: Optional[str] = None):
        self.url = url.replace("http", "ws").rstrip("/") + "/ws"
        self.token = token
        self.socket = None
        self.reconnect_interval = 3
        self.listeners: List[Callable[[Any], None]] = []
        self.is_connected = False
        self.current_filter = None
        self.pending_requests = {}

    async def connect(self):
        while True:
            try:
                async with websockets.connect(self.url) as websocket:
                    self.socket = websocket
                    self.is_connected = True
                    print("[ApexKit] Realtime Connected")

                    if self.current_filter:
                        await self.subscribe(self.current_filter)

                    async for message in websocket:
                        try:
                            msg = json.loads(message)
                            if 'request_id' in msg:
                                self._handle_request_response(msg)
                            else:
                                self.notify(msg)
                        except json.JSONDecodeError:
                            if message == "Pong":
                                continue
                            print(f"WS Parse Error: {message}")
            except Exception as e:
                self.is_connected = False
                print(f"[ApexKit] Disconnected: {e}. Retrying in {self.reconnect_interval}s...")
                await asyncio.sleep(self.reconnect_interval)

    async def subscribe(self, filter_data: Dict[str, Any]):
        self.current_filter = filter_data
        if self.socket and self.is_connected:
            await self.socket.send(json.dumps({
                "type": "Subscribe",
                "payload": {
                    "collection_id": filter_data.get("collectionId"),
                    "record_id": filter_data.get("recordId"),
                    "event_type": filter_data.get("eventType"),
                    "filter": filter_data.get("dataFilter"),
                    "channel": filter_data.get("channel"),
                    "custom_event": filter_data.get("customEvent")
                }
            }))

    async def send_signal(self, channel: str, event_name: str, data: Any):
        if self.socket and self.is_connected:
            await self.socket.send(json.dumps({
                "type": "Signal",
                "payload": {"channel": channel, "event": event_name, "data": data}
            }))

    async def search(self, collection_id: Any, query: str, limit: int = 10):
        if not self.is_connected or not self.socket:
            raise Exception("Socket not connected")

        request_id = str(uuid.uuid4())
        loop = asyncio.get_event_loop()
        future = loop.create_future()

        self.pending_requests[request_id] = future

        await self.socket.send(json.dumps({
            "type": "Search",
            "payload": {
                "collection_id": int(collection_id),
                "query": query,
                "limit": limit,
                "request_id": request_id
            }
        }))

        try:
            return await asyncio.wait_for(future, timeout=5.0)
        except asyncio.TimeoutError:
            self.pending_requests.pop(request_id, None)
            raise Exception("Search request timed out")

    def on_event(self, callback: Callable[[Any], None]):
        self.listeners.append(callback)
        return lambda: self.listeners.remove(callback)

    def notify(self, msg: Any):
        for listener in self.listeners:
            listener(msg)

    def _handle_request_response(self, msg: Any):
        request_id = msg.get('request_id')
        future = self.pending_requests.pop(request_id, None)
        if future:
            if msg.get('type') == "Error":
                future.set_exception(Exception(msg.get('message')))
            else:
                future.set_result(msg.get('results'))

class ApexKitRealtimeSSEClient:
    def __init__(self, base_url: str):
        self.base_url = base_url.rstrip("/")
        self.listeners: List[Callable[[Any], None]] = []

    def connect(self, channel: str = None, event_name: str = None):
        import sseclient # pip install sseclient-py
        params = {}
        if channel: params['channel'] = channel
        if event_name: params['event'] = event_name

        url = f"{self.base_url}/sse"
        response = requests.get(url, params=params, stream=True)
        client = sseclient.SSEClient(response)

        for event in client.events():
            try:
                msg = json.loads(event.data)
                self.notify(msg)
            except Exception as e:
                print(f"[ApexKit] SSE Parse Error: {e}")

    def on_event(self, callback: Callable[[Any], None]):
        self.listeners.append(callback)
        return lambda: self.listeners.remove(callback)

    def notify(self, msg: Any):
        for listener in self.listeners:
            listener(msg)
